data_pipeline/google_analytics/ga_pipeline.py

```diff
@@ -0,0 +1,180 @@
+    GoogleAnalyticsClient
+)
+
+LOGGER = logging.getLogger(__name__)
+
+GA_DATE_RANGE_KEY = "date_range"
+
+
+def transform_response_to_bq_compatible_record(
+        response
+) -> Iterable[dict]:
+    for report in response.get('reports', []):
+        column_header = report.get('columnHeader', {})
+        dimension_headers = column_header.get('dimensions', [])
+        metric_headers = column_header.get('metricHeader', {}).get(
+            'metricHeaderEntries', []
+        )
+
+        for row in report.get('data', {}).get('rows', []):
+            bq_json_formatted_record = {}
+            dimensions = row.get('dimensions', [])
+            date_range_values = row.get('metrics', [])
+
+            for header, dimension in zip(dimension_headers, dimensions):
+                bq_json_formatted_record[
+                    standardize_field_name(header)
+                ] = dimension
+            for i, values in enumerate(date_range_values):
+                bq_json_formatted_record[GA_DATE_RANGE_KEY] = str(i)
+                for metric_header, value in zip(
+                        metric_headers, values.get('values')
+                ):
+                    bq_json_formatted_record[
+                        standardize_field_name(metric_header.get('name'))
+                    ] = value
+
+            yield bq_json_formatted_record
+
+
+def etl_google_analytics(
+        ga_config: GoogleAnalyticsConfig,
+        start_date: datetime,
+        end_date: datetime = None,
+):
+    current_timestamp_as_string = get_current_timestamp_as_string()
+    analytics = GoogleAnalyticsClient()
+    end_date = end_date or datetime.now()
+    from_date = start_date.strftime("%Y-%m-%d")
+    to_date = end_date.strftime("%Y-%m-%d")
+    dimensions = [
+        {
+            'name': dim_name
+        } for dim_name in ga_config.dimensions
+    ]
+    metrics = [
+        {
+            'expression': metric_name
+        } for metric_name in ga_config.metrics
+    ]
+    page_token = None
+    while True:
+
+        response = analytics.get_report(
+            date_ranges=[{'startDate': from_date, 'endDate': to_date}],
+            view_id=ga_config.ga_view_id,
+            metrics=metrics,
+            dimensions=dimensions,
+            page_token=page_token
+        )
+        reports = response.get('reports', [])
+        # api can return multiple reports, however,
+        # this implementation considers single report
+        page_token = (
+            reports[0].get('nextPageToken')
+            if len(reports) == 1 else None
+        )
+
+        transformed_response_with_provenance = (
+            add_provenance(
+                transform_response_to_bq_compatible_record(
+                    response
+                ),
+                ga_config.import_timestamp_field_name,
+                current_timestamp_as_string,
+                ga_config.record_annotations
+            )
+        )
+        with NamedTemporaryFile() as temp_file:
```

Comment:
same as https://github.com/elifesciences/data-hub-ejp-xml-pipeline/pull/40/files#r433135539

Reply:
This comment is marked as resolved. Are we going to replace `NamedTemporaryFile`? I would really love to see it go :-)

Reply:
we need some temporary file that we can open for write and/or append, close, and re-open for read, and I can't just figure out how to use ordinary `TemporaryFile` for that, but `NameTemporaryFile`

Reply:
Okay, I think I may not explained properly what I am after. I fixed the link to the related conversation. I am suggesting to use `TemporaryDirectory` and then just create a file within that temporary directory.